File: backend/app/models/pitcher_ability.rb
class PitcherAbility < ApplicationRecord
      # RANKS = { G: 0, F: 1, E: 2, D: 3, C: 4, B: 5, A: 6, S: 7 }.freeze      
      RANKS = { g: 0, f: 1, e: 2, d: 3, c: 4, b: 5, a: 6, s: 7 }.freeze

      # バリデーション
      validates :pitch_velocity, inclusion: 0..180, allow_nil: true
      validates :control, inclusion: 0..100, allow_nil: true
      validates :stamina, inclusion: 0..100, allow_nil: true
      validates :w_risp, inclusion: { in: ['G','F','E','D','C','B','A','S'] }, allow_nil: true
      validates :heather, inclusion: { in: ['G','F','E','D','C','B','A','S'] }, allow_nil: true
      validates :vs_lbh, inclusion: { in: ['G','F','E','D','C','B','A','S'] }, allow_nil: true
      validates :agile, inclusion: { in: ['G','F','E','D','C','B','A','S'] }, allow_nil: true
      validates :poise, inclusion: { in: ['G','F','E','D','C','B','A','S'] }, allow_nil: true
      validates :grit, inclusion: { in: ['G','F','E','D','C','B','A','S'] }, allow_nil: true
      validates :recovery, inclusion: { in: ['G','F','E','D','C','B','A','S'] }, allow_nil: true

      validates :w_risp_rank, :heather_rank, :vs_lbh_rank, :agile_rank, :poise_rank, :grit_rank, :recovery_rank, presence: true

      enum w_risp_rank: RANKS, _prefix: true
      enum heather_rank: RANKS, _prefix: true
      enum vs_lbh_rank: RANKS, _prefix: true
      enum agile_rank:  RANKS, _prefix: true
      enum poise_rank: RANKS, _prefix: true
      enum grit_rank: RANKS, _prefix: true
      enum recovery_rank: RANKS, _prefix: true

      # アソシエーション
      belongs_to :player_season
end

File: frontend/src/models/pitcher_ability.ts
// ドメインモデル
export type PitcherAbility = {
  pitch_velocity: number | null;
  control: number | null;
  stamina: number | null;
  w_risp_rank: string | null;
  heather_rank: string | null;
  vs_lbh_rank: string | null;
  agile_rank: string | null;
  poise_rank: string | null;
  grit_rank: string | null;
  recovery_rank: string | null;
  special_ability?: string | null;
};

export const createPitcherAbility = (overrides?: Partial<PitcherAbility>): PitcherAbility => ({
  pitch_velocity: null,
  control: null,
  stamina: null,
  w_risp_rank: 'D',
  heather_rank: 'D',
  vs_lbh_rank: 'D',
  agile_rank: 'D',
  poise_rank: 'D',
  grit_rank: 'D',
  recovery_rank: 'D',
  special_ability: '',
  ...overrides,
});

// 送信フォーム用
export type PitcherAbilityForm = {
  pitch_velocity: string;
  control: string;
  stamina: string;
  w_risp_rank: string;
  heather_rank: string;
  vs_lbh_rank: string;
  agile_rank: string;
  poise_rank: string;
  grit_rank: string;
  recovery_rank: string;
  special_ability: string;
};

export const buildPitcherAbilityPayload = (
  form: PitcherAbilityForm
): PitcherAbility => ({
  pitch_velocity: form.pitch_velocity ? Number(form.pitch_velocity) : 0,
  control: form.control ? Number(form.control) : 0,
  stamina: form.stamina ? Number(form.stamina) : 0,
  w_risp_rank: form.w_risp_rank || 'D',
  heather_rank: form.heather_rank || 'D',
  vs_lbh_rank: form.vs_lbh_rank || 'D',
  agile_rank: form.agile_rank || 'D',
  poise_rank: form.poise_rank || 'D',
  grit_rank: form.grit_rank || 'D',
  recovery_rank: form.recovery_rank || 'D',
  special_ability: form.special_ability || null,
});